Drop StdCall in favor of Cdecl for x86.

JunielKatarn opened the pull request:
Being built independently, React Native Windows doesn't need to be built with the `__stdcall` convention.
Moreover, many third-party dependencies are developed assuming `__cdecl` is used (i.e. OpenSSL).

- Annotates all non-class-member (which use `__thiscall`) exported functions as `__cdecl`.
- Switches to a non-forked OpenSSL build.
- Resolves #4096.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/4115)

NikoAri commented on .gitmodules:
```diff
@@ -4,3 +4,8 @@
 	url = https://github.com/ReactWindows/yoga
+[submodule "folly"]
+	path = GitModules/folly
+	url = https://github.com/jurocha-ms/folly.git
```
>https://github.com/jurocha-ms/folly.git [](start = 7, length = 39)

Why is this in your personal repo?

JunielKatarn replied:
We need to fork to apply our patches, and we don't have a `microsoft/folly` repository.
This can be changed and fixed in the future trivially.
Moreover, if Facebook accepts this change, we can go back to referencing their upstream repo.

NikoAri commented on vnext/DeforkingPatches/ReactCommon/yoga/yoga/Utils.cpp:
```diff
@@ -0,0 +1,66 @@
+/*
```
Why are these files here? Is it because we removed dependency on yoga lib somewhere?

JunielKatarn replied:
These are just modified files to used to override the original Yoga sources.

We still depend on Yoga same way as before.

NikoAri commented on vnext/DeforkingPatches/ReactCommon/yoga/yoga/Yoga.h:
```diff
@@ -0,0 +1,368 @@
+    *YGCloneNodeFunc)(YGNodeRef oldNode, YGNodeRef owner, int childIndex);
+
+// YGNode
+WIN_EXPORT YGNodeRef __cdecl YGNodeNew(void);
```
>WIN_EXPORT [](start = 0, length = 10)

Would it make sense to insert "__cdecl" into existing WIN_EXPORT macro?

NikoAri commented on vnext/Desktop/packages.config:
```diff
@@ -4,5 +4,5 @@
   <package id="Microsoft.ChakraCore.vc140" version="1.11.13" targetFramework="native" developmentDependency="true" />
   <package id="Microsoft.Windows.CppWinRT" version="2.0.190730.2" targetFramework="native" />
-  <package id="ReactWindows.OpenSSL.StdCall.Static" version="1.0.2-p.2" targetFramework="native" />
+  <package id="ReactWindows.OpenSSL.v141.Static" version="1.1.1-d.1" targetFramework="native" />
```
>v141 [](start = 36, length = 4)

For my understanding, what is v141?

JunielKatarn replied:
`MSVC 14.1`.
Because we may be moving soon to `v142`, I want to keep this build specific and potentially generate a new one using `v142` when the change happens.

NikoAri commented on vnext/Desktop/packages.config:
```diff
@@ -4,5 +4,5 @@
   <package id="Microsoft.ChakraCore.vc140" version="1.11.13" targetFramework="native" developmentDependency="true" />
   <package id="Microsoft.Windows.CppWinRT" version="2.0.190730.2" targetFramework="native" />
-  <package id="ReactWindows.OpenSSL.StdCall.Static" version="1.0.2-p.2" targetFramework="native" />
+  <package id="ReactWindows.OpenSSL.v141.Static" version="1.1.1-d.1" targetFramework="native" />
```
>d [](start = 64, length = 1)

(nit) would it be possible to just have more common version format, eg:
1.2.3.4

JunielKatarn replied:
Because of OpenSSL's particular versioning scheme, this would not work for us.
Next minor release will be `1.1.1e`, but we also need to be able to version our nugets, for fixes and incremental changes based on the exact same source base.

Plus, this convention is compliant with Semantic Versioning ("semver").

Becaue nothing about OpenSSL or its package is exported by RNW, this shouldn't have transcending impact.

NikoAri commented on vnext/Desktop/WebSocket.cpp:
```diff
@@ -402,34 +402,34 @@ void BaseWebSocket<Protocol, SocketLayer, Stream, Resolver>::Ping() {
 
 template <typename Protocol, typename SocketLayer, typename Stream, typename Resolver>
-void BaseWebSocket<Protocol, SocketLayer, Stream, Resolver>::SetOnPing(function<void()> &&handler) {
+void BaseWebSocket<Protocol, SocketLayer, Stream, Resolver>::SetOnPing(function<void __cdecl()> &&handler) {
```
>__cdecl [](start = 85, length = 7)

hmm, is this needed? how does std::function deal with calling convention?

JunielKatarn replied:
I do not know it for a fact, but suspect that a consuming application built with `__stdcall` may have the functors act as `void __stdcall` functions. With this change, IWebSocket should expect functors that are equivalent to `void __cdecl`.

Am I making a wrong assumption here?
In any case, that change should be innocuous and could be reverted if I'm proven wrong.

JunielKatarn replied:
Confirmed. If a lambda is provided without expressing calling convention, it will go with the compiler's default.
It's best to require __cdecl explicitly.

NikoAri commented:
> void NativeTraceEventSource::UninitializeTracing(uint32_t cookie) {

Shouldn't we have __cdecl here too?

---
Refers to: vnext/Desktop/ABI/NativeTraceEventSource.cpp:59 in e452260. [](commit_id = e4522607e3fe7d149df4ede3c4317e3d8078c682, deletion_comment = False)

NikoAri commented:
Current change achieves its goal of standardizing naming convention. Comments to improve this are welcome. 

Windows uses "WINAPI" and "WINAPIV" for stdcall and cdecl respectively. Perhaps we could follow similar pattern.

---
In reply to: [360698414](https://github.com/microsoft/react-native-windows/pull/4115#issuecomment-360698414) [](ancestors = 360698414)

NikoAri commented on vnext/DeforkingPatches/ReactCommon/yoga/yoga/Yoga.h:
```diff
@@ -0,0 +1,368 @@
+    YGNodeRef child,
+    uint32_t index);
+
+WIN_EXPORT void __cdecl YGNodeRemoveChild(YGNodeRef node, YGNodeRef child);
```
>WIN_EXPORT [](start = 0, length = 10)

If we are already using "WIN_EXPORT" macro, perhaps we can use "WINAPIV" macro as well? So this line would be:
WIN_EXPORT void WINAPIV YGNodeRemoveChild(YGNodeRef node, YGNodeRef child);

JunielKatarn replied:
OK, point taken.
I'm reluctant to use Windows headers everywhere, especially if they are not needed for anything else, but it's fair to use in our patched Yoga sources.

I'll update accordingly.

kmelmon commented:
+1  can we discuss other options here?  I worry this is creating too much debt.

---
In reply to: [360698414](https://github.com/microsoft/react-native-windows/pull/4115#issuecomment-360698414) [](ancestors = 360698414)

JunielKatarn commented:
> Patching upstream source and manually marking everything Office uses with cdecl is quite fragile

True. That's why things are like they are right now, but we have ended up facing the same issue by keeping `__stdcall` as the absolute default. I'll ping you offline to elaborate, for the concerns are on dependent projects and not directly on RNW.

NickGerleman commented:
@NikoAri this does accomplish the goal, but in doing so, we:
- Need to maintain a forked copy of Facebook Yoga, that we touch every two months during upgrades
- Have a separate repo for folly we need to maintain (which should be attached to MSFT and have reviews instead of being personal)
- I believe we would also need to fork (and maintina) more React Native APIs in ReactCommon that we export, but this change doesn't seem to do that. E.g. the instance management stuff we call directly in devmain.
- Have some parts of code state calling convention explicitly but not others, where we will discover failures once in devmain.

The benefits I know of for this change would be to avoid having a patched OpenSSL, and to allow consuming third party binaries on x86. 

To me, this changes makes sense once we've moved the world to the Microsoft.ReactNative ABI-safe boundary to Facebook code. At that point we only need to have that part state its calling convention. (Though we still need to do something for Yoga) @vmoroz is working on this.

NickGerleman commented:
> True. That's why things are like they are right now, but we have ended up facing the same issue by keeping __stdcall as the absolute default. I'll ping you offline to elaborate, for the concerns are on dependent projects and not directly on RNW.

@JunielKatarn is this in reference to the OpenSSL break?

JunielKatarn commented:
> WINAPIV

That would require adding more dependencies on `Windows.h` than needed.
I believe we should not add this complexity.

JunielKatarn commented:
> > True. That's why things are like they are right now, but we have ended up facing the same issue by keeping __stdcall as the absolute default. I'll ping you offline to elaborate, for the concerns are on dependent projects and not directly on RNW.
> 
> @JunielKatarn is this in reference to the OpenSSL break?

Correct.
Also, OpenSSL is a sensitive dependency, so it's preferable to not patch it, even if other components need to be patched instead.

I'm aware of dependent software being potentially impacted.
I'll start testing these changes now.

NikoAri commented:
WINAPIV is declared inside of MinWinDef.h, not Windows.h

From quickly taking a look of contents of MinWinDef.h, it looks fairly self contained

---
In reply to: [587938026](https://github.com/microsoft/react-native-windows/pull/4115#issuecomment-587938026) [](ancestors = 587938026)

JunielKatarn commented:
> I believe we would also need to fork (and maintina) more React Native APIs in ReactCommon that we export, but this change doesn't seem to do that. E.g. the instance management stuff we call directly in devmain.

I will be testing this today.
Note, if these changes affect depending projects, we can partially revert this change by:
- Going back to `ReactWindows.OpenSSL.StdCall.Static`.
- Setting `CallingConvention` back to `StdCall`.

I'll merge this change after addressing feedback.

NickGerleman commented:
@junielkatarn apart from changing defs, do we need to change OpenSSL source at all, or just change compiler flags when producing the x86 binary?

I mentioned needing to patch more in ReactCommon, but as soon as we add all React, Yoga, and Folly sources exporting anything as patches, we're going back to near square one for Native deforking efforts, and people will almost definitely patch our code even note. I was just about to remove some of the last of the patches there. 

Given we were going to remove the rest, would you be willing to take efforts to merge during upgrades, or build tooling we'll need to detect when these are out of date when we're pulling from master? Without exaggeration, this change has the potential to add engineer-months of extra effort in the future relative to changing only OpenSSL.

Either way, I agree with @kmelmon that this change has large enough implications that we should discuss with everyone involved. Checkin should be blocked until this happens.

JunielKatarn commented:
> @JunielKatarn apart from changing defs, do we need to change OpenSSL source at all, or just change compiler flags when producing the x86 binary?

@NickGerleman With this change, OpenSSL can be built "off the shelf" with not a single change to sources or compiler flags.
Package `ReactWindows.OpenSSL.v141.Static` already reflects this.

NickGerleman commented:
I talked to @JunielKatarn offline. We shouldn't need to patch React Native functions, who are using "thiscall". Since there's not a sane way around this for OpenSSL this can make sense as a stopgap. @JunielKatarn will take on ownership of maintaining the patches during upgrades, and upstreaming them.

vmoroz commented on vnext/PropertySheets/x86.props:
```diff
@@ -9,7 +9,7 @@
   <ItemDefinitionGroup>
     <ClCompile>
-      <CallingConvention>StdCall</CallingConvention>
+      <CallingConvention>Cdecl</CallingConvention>
```
>Cdecl [](start = 25, length = 5)

If we do this, then do we need to add __cdecl?

JunielKatarn replied:
So we can export these headers and consuming projects are not forced to use `__cdecl`.

WillerZ commented:
Folly isn't designed to be a DLL: it doesn't have an ABI, or versions.

The current groupthink internally is that if some project wants to make a folly DLL for their own reasons then it's OK for them to have the burden of choosing the calling convention with the appropriate `/G` flag and making sure they are using the same one consistently in their project.

acoates-ms commented:
Agreed.  We should just wait for the work to remove folly from the API boundary before we switch the calling convention used inside react-native-windows back to normal.

JunielKatarn commented:
> Folly isn't designed to be a DLL: it doesn't have an ABI, or versions.
> 
> The current groupthink internally is that if some project wants to make a folly DLL for their own reasons then it's OK for them to have the burden of choosing the calling convention with the appropriate `/G` flag and making sure they are using the same one consistently in their project.

Problem is, we found a case where there are two projects (Say `A` which depends on `B`, and both depend on Folly, which is statically linked into `B`) that use different calling conventions.
Thus, whenever `A` calls into a Folly method, it will try to load it with a different calling convention, and we'll hit runtime crashes.

Annotating the proper calling convention would ensure whichever `/G` switch is used won't affect any consuming projects.
That said, if my team judges this as non-urgent, we can drop the subject for now.

WillerZ commented:
> we found a case where there are two projects (Say `A` which depends on `B`, and both depend on Folly, which is statically linked into `B`) that use different calling conventions.

The model we expect here is for both A and B to statically link to folly and that neither of them should be exporting any folly symbols.

JunielKatarn commented:
> > we found a case where there are two projects (Say `A` which depends on `B`, and both depend on Folly, which is statically linked into `B`) that use different calling conventions.
> 
> The model we expect here is for both A and B to statically link to folly and that neither of them should be exporting any folly symbols.

We'll look into that :)
